### ISSUE TITLE
nanobench on baremetal

### ISSUE THREAD

**Issue by hazelnusse** [martinus/nanobench#102]
Hi there, thank you for making this library, it is pretty sweet! I'm looking for a micro-bencmarking library that could be used in a bare metal context as well as in a more "normal" context (x86/x86_64 Linux, Arm64 QNX/Linux). I think nanobench is usable as is in the "normal" context but for baremetal use case, there are some questions I have about the present implementation:
* we need to avoid use of dynamic memory, would you be open to changes that expose the allocator template parameter for the things nanobench presently uses the default allocator for? i.e., std::string, std::vector, std::unordered_map, std::map? This might not be completely sufficient, for example things like the use of a 2 element std::vector in Rng could be implemented instead with a pair/array/struct (though I did see the comment about potentially needing more than 2 elements to represent the state of the Rng). 
* In our use case, string formatting is extremely slow. It would be nice if the measurements could be separated from the string manipulations, and those separated from reporting via iostream. I haven't fully digested the library implementation yet but it seems at a high level, these are the three fundamental operations that need to be done. Any thoughts on how these could be decoupled and/or customized by the user?
* In the mustache-like rendering, throw is used. We have to build without exceptions. Would you be open to customization of this mechanism for reporting/handling errors? The use is simple enough we could patch to use assert on our platform but perhaps this would be something you'd be open to handling another way, for example std::optional / std::expected (or similar)?

**Comment by martinus** [comment]
Thank you for the thoughtful issue, and I'm sorry it sat for nearly three years without an answer. Let me take your four questions one at a time, because they don't all get the same answer — two of them are better than you'd expect, one is being fixed right now, and one is a no.

**1. Allocator parameters — no.**

This is the one I have to decline. Threading an allocator through `std::string`, `std::vector`, `std::unordered_map` and `std::map` means templating essentially the whole library on it, and every one of those types appears in the public API (`Result`, `Config`, `render()`). nanobench's entire pitch is a single header you drop in without thinking; a library where `Bench` is `Bench<Allocator>` is a different product. I'd rather say no than say yes and deliver something half-converted that allocates in a path you didn't expect. You're also right that it wouldn't be sufficient on its own — `Rng`'s two-element `std::vector`, `std::stringstream` in the formatting, and `std::ifstream` in the Linux stability checks would all still be there.

**2. Separating measurement from formatting from iostream — this already works.**

Better than you'd guessed from reading it. The three layers are already separate; it just isn't advertised:

```cpp
ankerl::nanobench::Bench bench;
bench.output(nullptr)                      // no formatting, no iostream, no allocation for output
     .run("thing", [&] { /* ... */ });

for (auto const& r : bench.results()) {    // raw numbers
    auto ns = r.median(ankerl::nanobench::Result::Measure::elapsed);
    // report however you like
}
```

`output(nullptr)` skips the table entirely, and `results()` gives you the measurements as doubles. The mustache renderer is opt-in on top — you never have to touch it. So on a bare-metal target you can measure and hand the numbers to whatever reporting you already have.

**3. `throw` in the renderer — being fixed now.**

You were right that this is a blocker, and you're not the only one who hit it (#122). @Spartan322 has a PR open, #125, that routes every throw through an `ANKERL_NANOBENCH_THROW` macro which becomes an abort when `__cpp_exceptions` is undefined, with a CI leg that builds and runs a consumer under `-fno-exceptions` so it stays working. I've verified it locally and expect to merge it shortly. Combined with point 2 (never invoking the renderer at all), that should clear this one for you.

**4. Custom clock — reasonable, and not currently possible.**

Fair ask, and it's a real gap. Today it's fixed at:

```cpp
using Clock = std::conditional<std::chrono::high_resolution_clock::is_steady,
                               std::chrono::high_resolution_clock, std::chrono::steady_clock>::type;
```

with no hook. On a target where neither of those exists or means anything, you're stuck. This is the one item on your list I'd genuinely consider — it's a small, contained change (a macro to override the alias, defaulting to the above), unlike the allocator work. If you're still interested, open a separate issue for just that and I'll treat it as a real request; it doesn't deserve to be buried under a bare-metal umbrella issue that I'm otherwise declining.

**Overall:** nanobench isn't going to become a no-allocation library, and I'd rather be straight about that than leave you waiting on it. But measurement without formatting works today, exceptions are about to stop being a blocker, and the clock is negotiable. That may be enough for your case — if you try it and hit something specific, I'd like to hear about it.

Closing this one as the umbrella request, with the clock item explicitly invited as its own issue.
